src/trees/bp/mod.rs

```diff
@@ -0,0 +1,536 @@
+/// `BLOCK_SIZE` generic parameter.
+/// The block size should be chosen based on the expected size of the tree and the available memory.
+/// Smaller block sizes increase the size of the supporting data structure but reduce the time
+/// complexity of some operations by a constant amount.
+/// Very large block sizes are best combined with the `u16_lookup` feature to keep the query time
+/// low.
+///
+/// ## Unbalanced Parentheses
+/// The tree is implemented in a way to theoretically support unbalanced parenthesis expressions
+/// (which encode invalid trees) without panicking.
+/// However, some operations may behave erratically if the parenthesis expression isn't balanced.
+/// Generally, operations specify if they require a balanced tree.
+///
+/// The results of the operations are unspecified,
+/// meaning no guarantees are made about the stability of the results across versions
+/// (except the operations not panicking).
+/// However, for research purposes, this behavior can be useful and should yield expected results
+/// in most cases.
+///
+/// # Example
+/// ```rust
+/// # #![allow(long_running_const_eval)] // for some reason this is needed for test cases
+/// use vers_vecs::{BitVec, BpDfsBuilder, BpTree, DfsTreeBuilder, Tree};
+///
+/// let mut builder = BpDfsBuilder::<512>::new();
+///
+/// // build the tree by depth-first traversal
+/// builder.enter_node();
+/// builder.enter_node();
+/// builder.enter_node();
+/// builder.leave_node();
+/// builder.enter_node();
+/// builder.leave_node();
+/// builder.leave_node();
+/// builder.enter_node();
+/// builder.leave_node();
+/// builder.leave_node();
+///
+/// let tree = builder.build().unwrap();
+/// let root = tree.root().unwrap();
+/// assert_eq!(root, 0);
+/// assert_eq!(tree.first_child(root), Some(1));
+/// assert_eq!(tree.next_sibling(1), Some(7));
+/// assert_eq!(tree.next_sibling(7), None);
+///
+/// assert_eq!(root, 0);
+/// assert_eq!(tree.depth(2), 2);
+/// assert_eq!(tree.depth(7), 1);
+/// ```
+///
+/// [`RsVec`]: RsVec
+pub struct BpTree<const BLOCK_SIZE: usize = DEFAULT_BLOCK_SIZE> {
+    vec: RsVec,
+    min_max_tree: MinMaxTree,
+}
+
+impl<const BLOCK_SIZE: usize> BpTree<BLOCK_SIZE> {
+    /// Construct a new `BpTree` from a given bit vector.
+    #[must_use]
+    pub fn from_bit_vector(bv: BitVec) -> Self {
+        let min_max_tree = MinMaxTree::excess_tree(&bv, BLOCK_SIZE);
+        let vec = bv.into();
+        Self { vec, min_max_tree }
+    }
+
+    /// Search for a position where the excess relative to the starting `index` is `relative_excess`.
+    /// Returns `None` if no such position exists.
+    /// The initial position is never considered in the search.
+    /// Searches forward in the bit vector.
+    ///
+    /// # Arguments
+    /// - `index`: The starting index.
+    /// - `relative_excess`: The desired relative excess value.
+    pub fn fwd_search(&self, index: usize, mut relative_excess: i64) -> Option<usize> {
```

Comment:
Is it worthwhile making fwd_search and bwd_search public at all? Are there practical use cases for people to use these directly?